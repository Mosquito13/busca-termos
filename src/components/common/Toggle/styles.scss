@import '../../../sass/styles.scss';

.toggle {
  display: flex;
  align-items: center;

  &__input {
    width: 4.5rem;
    height: 2.5rem;
    border: 1px solid currentColor;
    border-radius: 50px;
    padding: 1px;
    display: flex;
    transition: background-color 0.2s;
    background-color: transparent;
    cursor: pointer;

    &--checked {
      @include themify() {
        background-color: themed('color-contrast');
      }
    }

    &--disabled {
      filter: brightness(0.6);
      cursor: not-allowed;
    }
  }

  &__circle {
    width: 2.1rem;
    height: 100%;
    background-color: currentColor;
    border-radius: 50%;
    transform: none;
    animation: uncheckToggle 0.2s linear 0s forwards;

    &--checked {
      animation: checkToggle 0.2s linear 0s forwards;
    }

    &--with-icon {
      border: 1px solid currentColor;
      transition: filter 0.2s;
      filter: contrast(0.3);
      display: flex;
      align-items: center;
    }

    &--with-icon#{&}--checked {
      filter: none;
    }
  }

  &__label {
    margin-left: 1rem;
    user-select: none;
    cursor: pointer;

    &--disabled {
      cursor: not-allowed;
      filter: brightness(0.6);
    }
  }
}
