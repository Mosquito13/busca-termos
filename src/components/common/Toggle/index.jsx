import classNames from 'classnames';

import './styles.scss';

const Toggle = ({ label, icon, value, onChange, disabled }) => {
  const classesInput = classNames(
    'toggle__input',
    value && 'toggle__input--checked',
    disabled && 'toggle__input--disabled'
  );

  const classesCircle = classNames(
    'toggle__circle',
    icon && 'toggle__circle--with-icon',
    value && 'toggle__circle--checked'
  );

  const classesLabel = classNames(
    'toggle__label',
    disabled && 'toggle__label--disabled'
  );

  const handleChange = onChange && !disabled ? onChange : null;

  const IconCmp = icon || '';

  return (
    <div className="toggle">
      <div className={classesInput} onClick={handleChange}>
        <div className={classesCircle}>{IconCmp}</div>
      </div>
      <div className={classesLabel} onClick={handleChange}>{label}</div>
    </div>
  );
};

export default Toggle;
