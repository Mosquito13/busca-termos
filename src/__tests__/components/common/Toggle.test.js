import { shallow } from 'enzyme';
import renderer from 'react-test-renderer';
import { FiActivity } from 'react-icons/fi';

import Toggle from '../../../components/common/Toggle';

describe(`${Toggle.name}`, () => {
  describe('tests', () => {
    let wrapper, onChangeSpy;

    beforeAll(() => {
      onChangeSpy = jest.fn();

      wrapper = shallow(<Toggle label="A label" onChange={onChangeSpy} />);
    });

    it('should render', () => {
      expect(wrapper.isEmptyRender()).toBeFalsy();
      expect(wrapper.find('.toggle')).toHaveLength(1);
    });

    it('should fire onChange', () => {
      wrapper.find('.toggle__input').simulate('click');
      expect(onChangeSpy).toBeCalledTimes(1);
    });

    it('should not fire onChange when it is disabled', () => {
      wrapper.setProps({ disabled: true });
      wrapper.find('.toggle__input').simulate('click');
      expect(onChangeSpy).toBeCalledTimes(0);
    });

    afterAll(() => {
      wrapper.unmount();
    });
  });

  describe('snapshots', () => {
    it('should render unchecked', () => {
      const tree = renderer.create(<Toggle value={false} />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render checked', () => {
      const tree = renderer.create(<Toggle value />).toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render with icon', () => {
      const tree = renderer
        .create(<Toggle value icon={<FiActivity />} />)
        .toJSON();

      expect(tree).toMatchSnapshot();
    });

    it('should render disabled', () => {
      const tree = renderer.create(<Toggle value disabled />).toJSON();

      expect(tree).toMatchSnapshot();
    });
  });
});
