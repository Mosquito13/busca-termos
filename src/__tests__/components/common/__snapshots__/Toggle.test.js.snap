// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toggle snapshots should render checked 1`] = `
<div
  className="toggle"
>
  <div
    className="toggle__input toggle__input--checked"
    onClick={null}
  >
    <div
      className="toggle__circle toggle__circle--checked"
    >
      
    </div>
  </div>
  <div
    className="toggle__label"
    onClick={null}
  />
</div>
`;

exports[`Toggle snapshots should render disabled 1`] = `
<div
  className="toggle"
>
  <div
    className="toggle__input toggle__input--checked toggle__input--disabled"
    onClick={null}
  >
    <div
      className="toggle__circle toggle__circle--checked"
    >
      
    </div>
  </div>
  <div
    className="toggle__label toggle__label--disabled"
    onClick={null}
  />
</div>
`;

exports[`Toggle snapshots should render unchecked 1`] = `
<div
  className="toggle"
>
  <div
    className="toggle__input"
    onClick={null}
  >
    <div
      className="toggle__circle"
    >
      
    </div>
  </div>
  <div
    className="toggle__label"
    onClick={null}
  />
</div>
`;

exports[`Toggle snapshots should render with icon 1`] = `
<div
  className="toggle"
>
  <div
    className="toggle__input toggle__input--checked"
    onClick={null}
  >
    <div
      className="toggle__circle toggle__circle--with-icon toggle__circle--checked"
    >
      <svg
        fill="none"
        height="1em"
        stroke="currentColor"
        strokeLinecap="round"
        strokeLinejoin="round"
        strokeWidth="2"
        style={
          Object {
            "color": undefined,
          }
        }
        viewBox="0 0 24 24"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <polyline
          points="22 12 18 12 15 21 9 3 6 12 2 12"
        />
      </svg>
    </div>
  </div>
  <div
    className="toggle__label"
    onClick={null}
  />
</div>
`;
